From d6d7eefd2de1c66860f508c9e5f9e7bfd0c44fd6 Mon Sep 17 00:00:00 2001
From: Ao Xu <ao.xu@amlogic.com>
Date: Fri, 24 Apr 2020 20:26:27 +0800
Subject: [PATCH 3/4] uvm: fix uvm secure buffer map crash issue [1/1]

PD#SWPL-24746

Problem:
for uvm secure buffer, cpu map will crash

Solution:
cpu do not map for uvm secure buffer

Verify:
u212

Change-Id: I8778c7baba4b855b92b56446232075801e97b455
Signed-off-by: Ao Xu <ao.xu@amlogic.com>
---
 drivers/amlogic/drm/meson_gem.c                   | 4 ++++
 drivers/amlogic/media/common/uvm/meson_uvm_core.c | 7 +++++++
 include/linux/amlogic/meson_uvm_core.h            | 1 +
 3 files changed, 12 insertions(+)

diff --git a/drivers/amlogic/drm/meson_gem.c b/drivers/amlogic/drm/meson_gem.c
index 4b28103..e98dff8 100644
--- a/drivers/amlogic/drm/meson_gem.c
+++ b/drivers/amlogic/drm/meson_gem.c
@@ -625,6 +625,10 @@ struct dma_buf *am_meson_drm_gem_prime_export(struct drm_device *dev,
 
 			if (meson_gem_obj->is_afbc)
 				info.flags |= BIT(UVM_FAKE_ALLOC);
+
+			if (meson_gem_obj->is_secure)
+				info.flags |= BIT(UVM_SECURE_ALLOC);
+
 			info.obj = &meson_gem_obj->ubo;
 			info.free = am_meson_drm_gem_unref_uvm;
 			dmabuf_bind_uvm_alloc(dmabuf, &info);
diff --git a/drivers/amlogic/media/common/uvm/meson_uvm_core.c b/drivers/amlogic/media/common/uvm/meson_uvm_core.c
index 0611d8b..6858a8e 100644
--- a/drivers/amlogic/media/common/uvm/meson_uvm_core.c
+++ b/drivers/amlogic/media/common/uvm/meson_uvm_core.c
@@ -86,6 +86,9 @@ static struct sg_table *meson_uvm_map_dma_buf(
 
 	UVM_PRINTK(1, "meson_uvm_map_dma_buf called, %s.\n", current->comm);
 
+	if (ua->flags & BIT(UVM_SECURE_ALLOC))
+		return sgt;
+
 	if (ua->flags & BIT(UVM_FAKE_ALLOC)) {
 		dma_map_page(attachment->dev, sg_page(sgt->sgl), 0,
 				UVM_FAKE_SIZE, direction);
@@ -111,6 +114,10 @@ static void meson_uvm_unmap_dma_buf(struct dma_buf_attachment *attachment,
 	ua = handle->ua;
 
 	UVM_PRINTK(1, "meson_uvm_unmap_dma_buf called, %s.\n", current->comm);
+
+	if (ua->flags & BIT(UVM_SECURE_ALLOC))
+		return;
+
 	if (ua->flags & BIT(UVM_FAKE_ALLOC))
 		dma_unmap_page(attachment->dev, sgt->sgl->dma_address,
 			UVM_FAKE_SIZE, direction);
diff --git a/include/linux/amlogic/meson_uvm_core.h b/include/linux/amlogic/meson_uvm_core.h
index 2750246..83ec490 100644
--- a/include/linux/amlogic/meson_uvm_core.h
+++ b/include/linux/amlogic/meson_uvm_core.h
@@ -38,6 +38,7 @@ enum uvm_alloc_flag {
 	UVM_IMM_ALLOC,
 	UVM_DELAY_ALLOC,
 	UVM_FAKE_ALLOC,
+	UVM_SECURE_ALLOC,
 };
 
 /**
-- 
2.7.4

